source/common/upstream/wrsq_scheduler.h

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "envoy/common/random_generator.h"
+#include "envoy/upstream/scheduler.h"
+
+#include "common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// Weighted Random Selection Queue (WRSQ) Scheduler
+// ------------------------------------------------
+// This scheduler keeps a queue for each unique weight among all objects inserted and adds the
+// objects to their respective queue based on weight. When performing a pick operation, a queue is
+// selected and an object is pulled. Each queue gets its own selection probability which is weighted
+// as the sum of all weights of objects contained within. Once a queue is picked, you can simply
+// pull from the top and honor the expected selection probability of each object.
+//
+// Adding an object will cause the scheduler to rebuild internal structures on the first pick that
+// follows. This operation will be linear on the number of unique weights among objects inserted.
+// Outside of this case, object picking is logarithmic with the number of unique weights. Adding
+// objects is always constant time.
+//
+// For the case where all object weights are the same, WRSQ behaves identical to vanilla
+// round-robin. If all object weights are different, it behaves identical to weighted random
+// selection.
+//
+// NOTE: This class only supports integral weights and does not allow for the changing of object
+// weights on the fly.
```

Comment:
> Also, I'm not sure how much we benefit from using floating-point weights. If we move forward with this scheduler, we should strongly consider switching to integral weights for the purpose of efficiency. The upside of floats is the dynamic range, so you can have weights of 10E1 and 10E10. I'm not sure that the relative difference between weights of 5.0, 5.00001, and 5.00002 really matters though.

I think you're onto something here. I was poking around the EDF scheduler and load balancer impl and can't find a good reason to keep the weights a double in the scheduler interface. The LbEndpoint weights are uint32s and the weights that the Host objects have are also uint32s.

I think it's worth cooking up a patch and then rebasing to address the Scheduler interface issue. Hopefully it's a straightforward change.

Reply:
I'm not sure that's sufficient. With the existing implementation, weights of 0.0000001 and 0.0000002 still result in a 1:2 balancing ratio. We'd have to re-normalize within the scheduler any time a weight is added or changed. If we want to use integer representations, we'd need to expose that at the function interface.

Reply:
Well, currently the only way to get weights like that is for it to be adjusted via LRLB or something like slow-start in https://github.com/envoyproxy/envoy/pull/13176. There would need to be not only a weight of 1 specified for a host in the proto, but also O(tens-of-thousands) of outstanding requests to get down to a number like that that the scaling wouldn't be able to represent. 

If for some reason we did end up with those numbers, they would just round down to zero until whatever calculation doing the adjustments results in something > 1e-5. I still think this can work.

Reply:
TBH I am not at all familiar with how the scheduler is actually used. If you say this will work in practice, I believe you. I'm arguing that this implementation doesn't match the Scheduler interface as proposed. If we want a minimum value, let's encode that either in a comment or (preferably) in the type of the weights.

Reply:
I think what I'm asking for to switch to fixed-point values. That allows us to encode in the type that, say with 2 decimal points of precision, 3.112 and 3.114 are sufficiently close as to be indistinguishable. Using integers as weights would just be a special case where # decimals = 0.

Reply:
I just caught your update here, sorry for the delay. Introducing this to the base scheduler type makes sense to me.

There are some other things I'd like to knock out before getting back to this, but I'll knock it out and update.

Reply:
I started doing this, but it quickly became a spookier change than I'd like to make across all the LB code. I encoded it in a comment instead, but if it makes folks wince we can chat this out a bit more and figure out a better path forward.

Reply:
Okay, ignoring my comments regarding integral weights, we really should consider changing the interface if this implementation isn't going to have the ability to update object weights.